network/transport/stcp_socket.c

```diff
@@ -0,0 +1,258 @@
+	sock_param.init_param.tcp = NULL;
+	ret = socket_init(&ldesc->sock, &sock_param);
+	if (IS_ERR_VALUE(ret)) {
+		ldesc->sock = NULL;
+		goto exit;
+	}
+
+	ret = trng_init(&ldesc->trng, param->trng_init_param);
+	if (IS_ERR_VALUE(ret)) {
+		ldesc->trng = NULL;
+		goto exit;
+	}
+
+	/* Set default configuration: TLS client socket */
+	ret = mbedtls_ssl_config_defaults(ldesc->conf,
+					  MBEDTLS_SSL_IS_CLIENT,
+					  MBEDTLS_SSL_TRANSPORT_STREAM,
+					  MBEDTLS_SSL_PRESET_DEFAULT);
+	if (IS_ERR_VALUE(ret))
+		goto exit;
+
+	/* Certificate validation is set as mandatory */
+	mbedtls_ssl_conf_authmode(ldesc->conf, MBEDTLS_SSL_VERIFY_NONE);
+
+	/* Decode the root certificate */
+	ret = mbedtls_x509_crt_parse(ldesc->ca_cert,
+				     (const unsigned char *)param->ca_cert,
+				     strlen(param->ca_cert) + 1);
+	if (IS_ERR_VALUE(ret))
+		goto exit;
+	/* Set issuer certificate authority. NULL for revoked certificates */
+	mbedtls_ssl_conf_ca_chain(ldesc->conf, ldesc->ca_cert, NULL);
+
+	/* Decode client certificate */
+	if (param->client_cert) {
+		ret = mbedtls_x509_crt_parse(ldesc->client_cert,
+					     (const unsigned char *)param->client_cert,
+					     strlen(param->client_cert) + 1);
+		if (IS_ERR_VALUE(ret))
+			goto exit;
+
+		if (param->client_key) {
+			ret = mbedtls_pk_parse_key(ldesc->client_pkey,
+						   (const unsigned char *)param->client_key,
+						   strlen(param->client_key) + 1,
+						   NULL, 0);
+			if (IS_ERR_VALUE(ret))
+				goto exit;
+		}
+		/* Set own certificate */
+		mbedtls_ssl_conf_own_cert(ldesc->conf, ldesc->client_cert,
+					  ldesc->client_pkey);
+	}
+
+
+	/* Config Random number generator */
+	mbedtls_ssl_conf_rng(ldesc->conf,
+			     (int (*)(void *, unsigned char *, size_t))
+			     trng_fill_buffer,
+			     (void *)ldesc->trng);
+
+	/* Set the resulting protocol configuration */
+	ret = mbedtls_ssl_setup(ldesc->ssl, ldesc->conf);
+	if (IS_ERR_VALUE(ret))
+		goto exit;
+
+	if (param->hostname) {
+		ret = mbedtls_ssl_set_hostname(ldesc->ssl, param->hostname);
+		if (IS_ERR_VALUE(ret))
+			goto exit;
+	}
+
+	/* Set socket callbacks */
+	mbedtls_ssl_set_bio(ldesc->ssl, ldesc->sock,
+			    (mbedtls_ssl_send_t *)socket_send,
+			    (mbedtls_ssl_recv_t *)tls_net_recv, NULL);
+
+	*desc = ldesc;
+
+	return SUCCESS;
+
+exit:
+	stcp_socket_remove(ldesc);
+
+	return ret;
+}
+
+/** @brief See \ref socket_remove */
+static int32_t stcp_socket_remove(struct stcp_socket_desc *desc)
+{
+	if (!desc)
+		return FAILURE;
+
+	stcp_socket_disconnect(desc);
+
+	mbedtls_x509_crt_free(desc->ca_cert);
+	mbedtls_x509_crt_free(desc->client_cert);
+	mbedtls_pk_free(desc->client_pkey);
+	mbedtls_ssl_config_free(desc->conf);
+	if (desc->trng)
+		trng_remove(desc->trng);
+	if (desc->sock)
+		socket_remove(desc->sock);
+	free(desc);
+
+	return SUCCESS;
+}
+
+/** @brief See \ref socket_connect */
+static int32_t stcp_socket_connect(struct stcp_socket_desc *desc,
+				   struct socket_address *addr)
+{
+	int32_t ret;
+
```

Comment:
shouldn't we check here for `desc` / `addr` if NULL?

Reply:
check for !desc.
Address is not dereferenced here so it is not a problem. The check is done in the next level anyway